### DISCUSSION TITLE
Should I fetch data from database in mount method or render method?

### DISCUSSION THREAD

**Discussion by phuclh** [livewire/livewire#2122] [1. Help]
I have 2 components:

- Websites: To fetch all websites (~300 websites) from the database.
- DetailWebsite: To display detailed info of a website when a user selects it on the list.

I did 2 tests:

In the first test, I fetched 300 websites in the `mount` method and assign the result to `public $websites`. Then I click a website on the list to view detail, it takes ~2s to reload the `Websites` component then it loads the `DetailWebsite` component after that.

```
public function mount()
{
    $this->fetchWebsites();
}

public function fetchWebsites() {.....}
```

<img width="364" alt="Screen Shot 2020-11-29 at 9 37 18 PM" src="https://user-images.githubusercontent.com/6707194/100572733-1808b980-328b-11eb-933d-e165952c8e11.png">

But in the 2nd test, I fetched data and pass to the view in `render` method.

```
public function render()
{
    return view('livewire.guest-posts', ['websites' => $this->fetchWebsites()]);
}
```
It only takes 800ms-1.5s (The time will have a big difference if I increase the items in the database). Why it can be faster than the 1st test since the `mount` method only runs one time at initial while render will be run every time I click a website to view detail.

This is my function to get the website detail:

```
public function selectWebsite($websiteId)
{
    $this->selectedWebsiteId = $websiteId;
    $this->emitTo('detail-website', 'findWebsite', $websiteId);
}
```

**Comment by joshhanley** [comment]
@phuclh you will probably find that the render method is quicker in that instance because the $websites data is only used inside the blade view (hence passing it directly to the view).

Where as if you use the mount method and assign the list to `$this->websites`, then when the component is finished rendering, it then serialises all public properties and makes them available to the Livewire front end.

So it all comes down to how you want to interact with your data as to whether to put it in mount or render.

By assigning it to a public property allows the data to be accessed on the front end and using Alpine, which could be of benefit in some instances but not necessary in others.

You may also want to access the list using bound input fields, in which case you would need to assign it to a public property. But if it's a datatable then it will work fine in render.

Let me know if there is anything you would like clarified!
Hope this helps!

**Reply by phuclh** [reply]
Thank you :D. So the one that makes it slower is Livewire has to serialize all public $property, right?

**Reply by joshhanley** [reply]
@phuclh no worries. Yes, most likely, especially if you've loaded a lot of data.
